src/rust/engine/engine_cffi/src/lib.rs

```diff
@@ -412,35 +443,11 @@ pub extern "C" fn scheduler_metrics(
         .flat_map(|(metric, value)| vec![externs::store_utf8(metric), externs::store_i64(value)])
         .collect::<Vec<_>>();
       if session.should_record_zipkin_spans() {
-        let workunits = session
-          .workunit_store()
-          .get_workunits()
-          .lock()
-          .iter()
-          .map(|workunit| {
-            let mut workunit_zipkin_trace_info = vec![
-              externs::store_utf8("name"),
-              externs::store_utf8(&workunit.name),
-              externs::store_utf8("start_secs"),
-              externs::store_u64(workunit.time_span.start.secs),
-              externs::store_utf8("start_nanos"),
-              externs::store_u64(u64::from(workunit.time_span.start.nanos)),
-              externs::store_utf8("duration_secs"),
-              externs::store_u64(workunit.time_span.duration.secs),
-              externs::store_utf8("duration_nanos"),
-              externs::store_u64(u64::from(workunit.time_span.duration.nanos)),
-              externs::store_utf8("span_id"),
-              externs::store_utf8(&workunit.span_id),
-            ];
-            if let Some(parent_id) = &workunit.parent_id {
-              workunit_zipkin_trace_info.push(externs::store_utf8("parent_id"));
-              workunit_zipkin_trace_info.push(externs::store_utf8(parent_id));
-            }
-            externs::store_dict(&workunit_zipkin_trace_info)
-          })
-          .collect::<Vec<_>>();
+        let workunits = session.workunit_store().get_workunits();
+
+        let value = workunits_to_py_tuple_value(&workunits.lock());
```

Comment:
I like this version more too now, also just because I feel like the the `.lock()` call is easier to see explicitly here, and that feels less surprising to me. I tend to write rust code in long `.map()`/`.flat_map()` chains as well, which sometimes gets more difficult to read.